contracts/Relay.sol

```diff
@@ -0,0 +1,284 @@
+    // Ecdsa.recover can recover the signer’s address. 
+    // If the signer is matched "isRelayer", it will be counted as a valid signature 
+    // and all signatures will be restored. 
+    // If the number of qualified signers is greater than Equal to threshold, 
+    // the verification is considered successful, otherwise it fails
+    function _checkSignature(
+        bytes32 hash,
+        bytes memory message,
+        bytes[] memory signatures
+    ) internal view returns (bool) {
+        require(
+            keccak256(message) == hash,
+            "Relay: The message does not match the hash"
+        );
+        require(signatures.length < 0xffffffff, "Relay: overflow");
+
+        uint16 count;
+        for (uint16 i = 0; i < signatures.length; i++) {
+            address signer = ECDSA.recover(ECDSA.toEthSignedMessageHash(hash), signatures[i]);
+            if (_isRelayer(signer)) {
+                count++;
+            }
+        }
+
+        uint8 threshold = uint8(
+            SafeMath.div(SafeMath.mul(uint256(count), 100), _getRelayerCount())
+        );
+
+        return threshold >= _getRelayerThreshold();
+    }
+
+    function _checkNetworkPrefix(bytes memory prefix) view public returns (bool) {
+      return assertBytesEq(_getNetworkPrefix(), prefix);
+    }
+
+    function updateRelayer(
+        bytes32 hash,
+        bytes memory message,
+        bytes[] memory signatures,
+        bytes32 benefit
+    ) public whenNotPaused {
+        // verify hash, signatures (The number of signers must be greater than 2/3 of the total)
+        require(
+            _checkSignature(hash, message, signatures),
+            "Relay: Bad relayer signature"
+        );
+
+        // decode message, check nonce and relayer
+        Input.Data memory data = Input.from(message);
+        (bytes memory prefix, uint32 nonce, address[] memory authorities) = Scale.decodeAuthorities(
+            data
+        );
+
+        require(_checkNetworkPrefix(prefix), "Relay: Bad network prefix");
+
+        // update nonce,relayer
+        _setRelayer(nonce, authorities, benefit);
+    }
+
+    function appendRoot(
+        bytes32 hash,
+        bytes memory message,
+        bytes[] memory signatures
+    ) public whenNotPaused {
+        // verify hash, signatures
+        require(
+            _checkSignature(hash, message, signatures),
+            "Relay: Bad relayer signature"
+        );
+
+        // decode message, check nonce and relayer
+        Input.Data memory data = Input.from(message);
+        (bytes memory prefix, uint32 index, bytes32 root) = Scale.decodeMMRRoot(data);
+
+        require(_checkNetworkPrefix(prefix), "Relay: Bad network prefix");
+
+        // append index, root
+        _appendRoot(index, root);
+    }
+
+    function resetRoot(uint32 index, bytes32 root) public onlyOwner {
+        _setRoot(index, root);
+        emit ResetRootEvent(_msgSender(), root, index);
+    }
+
+    function verifyAndDecodeReceipt(
```

Comment:
```suggestion
    function verifyMMRAndDecodeReceipt(
```

Reply:
Has been change to "verifyRootAndDecodeReceipt"